### PR TITLE
fix: Make sure to use backward-compatible names for resources created by the operator [RHIDP-2432]

### PR DESCRIPTION
**Pull request by rm3l**
## Description

This fixes a critical issue where, when upgrading existing resources using an operator upgraded from 1.1 to 1.2 (`main` branch), the operator would create and use a brand new database StatefulSet. This would cause the upgraded Backstage application to boot using an empty database, rather than starting with the existing database (which could already contain some data).

It also fixes a more general issue where resource naming was not backward-compatible, which would create for example 2 different ConfigMaps after an upgrade: one named `backstage-dynamic-plugins-<cr>` (created by the operator on the `main` branch) besides an already existing one named `<cr>-dynamic-plugins` (created in 1.1), which could create some confusion when looking at the resources in the namespace.

## Which issue(s) does this PR fix or relate to

- Fixes https://issues.redhat.com/browse/RHIDP-2432

## PR acceptance criteria

- [x] Tests
- [ ] Documentation
- [ ] If the bundle manifests have been updated, make sure to review the [`rhdh-operator.csv.yaml`](../.rhdh/bundle/manifests/rhdh-operator.csv.yaml) file accordingly

## How to test changes / Special notes to the reviewer

See the repro steps depicted in https://issues.redhat.com/browse/RHIDP-2432 and run the tests with the following command:
```
make test
make integration-test \
  ARGS=-focus='creates runtime object using raw configuration then updates StatefulSet to replace some immutable fields' \
  USE_EXISTING_CLUSTER=true
```

### CODE REVIEW

File: integration_tests/db_test.go
Comment (gazarenkov) [inline integration_tests/db_test.go:325]: Not sure what do we test with such a complex procedure...
It looks like that K8s correctly uses object type/name as a key and does not recreate it? Is it the point?

File: pkg/utils/utils.go
Comment (gazarenkov) [inline pkg/utils/utils.go:54]: I would really like to have it unified, as discussed and not to make us remember what is the pattern for each object type.  I do not think it impacts on functionality, the only drawback is that we leave old objects which can be solved by:
- upgrade documentation (recommending to remove it manually when upgrade from 1.1)
- additional ckeck/remove  (but I'd avoid it).

**Review by gazarenkov — CHANGES_REQUESTED** [review]
I would not like to have different name patterns for generated objects forever, let's unify it sooner than later.

File: controllers/backstage_controller.go
Comment (rm3l) [reply controllers/backstage_controller.go:184]: I find it more readable as it allows for a better understanding of the intent and exit conditions rather than long statements.
But also, as generally practiced, trying to keep the normal code path at a minimal indentation.
Anyway, I've tried to simplify this - see [`ec43e69` (#369)](https://github.com/janus-idp/operator/pull/369/commits/ec43e699618841fa0e68d42dd8523e6966a487a0)

File: pkg/utils/utils.go
Comment (rm3l) [reply pkg/utils/utils.go:54]: > I would really like to have it unified, as discussed and not to make us remember what is the pattern for each object type. I do not think it impacts on functionality
>
Well, in some cases, it impacted functionally. For example, I had to revert the DB Secret name to make sure a new Secret was not being created. Otherwise, connection to the existing DB would just fail.
I agree with you (at least in the medium/long run). But at the same time, found it really confusing to have both old and new objects in the cluster (without anything conveying which ones are deprecated).
I remember from a discussion someone edited the `<cr_name>-dynamic-plugins` ConfigMap (name in 1.1) to update their list of dynamic plugins. How would we handle this now that the name has changed to `backstage-dynamic-plugins-<cr_name>` in 1.2?

This is not ideal, but for the upcoming 1.2, I think it is safer to keep the existing naming to minimize the risk of additional bugs.

image

> the only drawback is that we leave old objects which can be solved by:
> 
> * upgrade documentation (recommending to remove it manually when upgrade from 1.1)
> * additional ckeck/remove  (but I'd avoid it).

The problem I foresee is with the Automatic Upgrade strategy with OLM, which, in my understanding, is much appreciated by users. IMO, we cannot claim backward compatibility and expect CR users to perform manual cleanup operations after an upgrade of the application (potentially automatic).
Or maybe a K8s Job (created by the operator) that does this migration automatically might make sense.
But that would be a separate issue.

File: integration_tests/db_test.go
Comment (gazarenkov) [reply integration_tests/db_test.go:325]: Sorry, still a bit confusing to me :) 

It sounds like you want to check 2 things:
1. the names of objects created by operator are expected
2. statefulset "patching" in a case when we try to change immutable fields (serviceName?) works as expected (delete/create)

If that's the case I believe have 2 simpler tests. First as unit test (just compare your model with predictable object names), second as integration test where you create CR and then update it (using raw config apparently) with changed SS. 
Does it make sense?

File: controllers/backstage_controller.go
Comment (gazarenkov) [reply controllers/backstage_controller.go:184]: Well, TBH I am not sure that this kind of "patching" is 100% correct (indeed there is a reason to keep some SS's fields immutable?) so I 'd like this patch looks as "isolated" as possible and would not look like the main flow.
Apparently we can not avoid it for upgrading but I would consider it as a rather temporary TBH.

File: pkg/utils/utils.go
Comment (gazarenkov) [reply pkg/utils/utils.go:54]: To make it clear

What is important to me is that the names of objects generated by Operator have unified pattern.
I.e. (as discussed) backstage-* or backstage-psql-* shape.
It is important for me that all of auto-generated items looks really as auto-generated, it also will simplify support.
And I think the sooner we do it the better (i.e. if we do not do it now it is a big chance it will live forever).

For your notes:

I think DB secret should not be affected, as I can see 1.1 name is secret/backstage-psql-secret-bs1 which fits the pattern, no?

For dynamic-plugins, I am not sure I understand the problem.
If you mean user change operator-config's configmap or secret - I do not think it will work anyway since it will be changed to desired state (from  operator-config). Or you meant something else? 

So far I can not see any backward compatibility problem.
I do not think existence of "old" objects can be taken as incompatible thing. 
Yes, we need to mention it in the Release Notes and recommend to clean it (I would rather not spend time for dedicated tooling for it).

File: controllers/backstage_controller.go
Comment (rm3l) [reply controllers/backstage_controller.go:184]: I took a quick look at how more specialized DB Operators would reconcile StatefulSets, and they appear to handle this the same way. See:
- https://github.com/zalando/postgres-operator/blob/master/pkg/cluster/resources.go#L198-L238
- https://github.com/CrunchyData/postgres-operator/blob/master/internal/controller/postgrescluster/pgadmin.go#L330-L351
- https://github.com/percona/percona-postgresql-operator/blob/main/internal/controller/postgrescluster/pgadmin.go#L332-L353

So, unless I've missed something, it seems to me that would be part of the "normal" code path.

File: pkg/utils/utils.go
Comment (rm3l) [reply pkg/utils/utils.go:54]: What pushes me to stick to the previous naming in 1.1 is because this is the second issue caused by this naming change (first one was https://github.com/janus-idp/operator/issues/355). And I wanted to prevent further issues due to this.
But alright, let's go ahead with your suggestions. I'll update the release notes text on the JIRA issue accordingly.
Thanks.

**Comment by rm3l** [comment]
@gazarenkov Thanks for the review comments. Apart from the one about the integration tests (which I think make sense as a way to guarantee the behavior and prevent further regression on this), I think it should be good. PTAL and approve if it is okay for you.

File: controllers/backstage_controller.go
Comment (gazarenkov) [reply controllers/backstage_controller.go:184]: OK
My proposal was mostly about to move this logic from to patchObject function since it is patch specific, nothing critically important.

File: integration_tests/db_test.go
Comment (gazarenkov) [reply integration_tests/db_test.go:325]: I think the tests showl be as focused as possible. It is still not clear to what does this test check. 

The way I see it is that you try to cover 2 cases: 
- the names of objects we create are expected. I.e:
backstage-<cr-name>,  not <cr-name>-backstage
backstage-psql-<cr-name>,  not backstage-db-<cr-name>
etc.
This one can be covered by simple unit test IMO?
- statefulset "patching" in a case when we try to change immutable fields (serviceName?) works as expected (delete/create), i.e:
the (integration) test can create CR with serviceName expected by 1.1 using raw configuration 
then it can update CR with default serviceName and validate it.
WDYT?

**Review by gazarenkov — COMMENTED** [review]
I'd like to see more targeted tests as mentioned in the comments, otherwise I doubt we'll be able to figure out what it was testing any time soon.

**Review by gazarenkov — COMMENTED** [review]
I'd like to see more targeted tests as mentioned in the comments, otherwise I doubt we'll be able to figure out what it was testing any time soon.

File: integration_tests/db_test.go
Comment (rm3l) [reply integration_tests/db_test.go]: Okay, I've tried to change the tests with this trade-off. See [`4758c86` (#369)](https://github.com/janus-idp/operator/pull/369/commits/4758c86e5758f2e294db3c7d8aa0ea31dd6cc8bd)

File: integration_tests/default-config_test.go
Comment (gazarenkov) [inline integration_tests/default-config_test.go:219]: it will be backstage-psql-<backstageName> after https://github.com/janus-idp/operator/pull/371

File: integration_tests/default-config_test.go
Comment (gazarenkov) [inline integration_tests/default-config_test.go:195]: Up to you, but, why do you need to update raw config?
IMO it is enough to just update Backstage with default config (especially after https://github.com/janus-idp/operator/pull/371 ) no?

Comment (rm3l) [reply integration_tests/default-config_test.go:195]: I wanted to check that something had actually changed to ensure the StatefulSet replacement. So it should be a field that is not overwritten upon the first reconciliation.
But with #371 in mind, I'll change that to another immutable field and update Backstage to the default config. Thanks.

Comment (rm3l) [reply integration_tests/default-config_test.go]: Updated in [`0f194b2` (#369)](https://github.com/janus-idp/operator/pull/369/commits/0f194b28ebcb5f9374d96c8e3ba781484c23fd5b). PTAL.

**Comment by rm3l** [comment]
/cherry-pick 1.2.x

But putting on hold until https://github.com/janus-idp/operator/pull/376 is merged, to avoid the change to the CSV.

/hold

**Review by gazarenkov — APPROVED** [review]

**Comment by rm3l** [comment]
/hold cancel

Rebased onto `main`, now that https://github.com/janus-idp/operator/pull/376 has been merged.

**Review by gazarenkov — APPROVED** [review]

**Pull request merged**
